File: crates/crates_io_database/src/models/trustpub/github_config.rs
use crate::schema::trustpub_configs_github;
use chrono::{DateTime, Utc};
use diesel::prelude::*;
use diesel_async::{AsyncPgConnection, RunQueryDsl};
use serde::Serialize;

#[derive(Debug, Identifiable, HasQuery, Serialize)]
#[diesel(table_name = trustpub_configs_github)]
pub struct GitHubConfig {
    pub id: i32,
    pub created_at: DateTime<Utc>,
    pub crate_id: i32,
    pub repository_owner: String,
    pub repository_owner_id: i32,
    pub repository_name: String,
    pub workflow_filename: String,
    pub environment: Option<String>,
}

impl GitHubConfig {
    pub async fn count_for_crate(conn: &mut AsyncPgConnection, crate_id: i32) -> QueryResult<i64> {
        trustpub_configs_github::table
            .filter(trustpub_configs_github::crate_id.eq(crate_id))
            .count()
            .get_result(conn)
            .await
    }
}

#[derive(Debug, Insertable)]
#[diesel(table_name = trustpub_configs_github, check_for_backend(diesel::pg::Pg))]
pub struct NewGitHubConfig<'a> {
    pub crate_id: i32,
    pub repository_owner: &'a str,
    pub repository_owner_id: i32,
    pub repository_name: &'a str,
    pub workflow_filename: &'a str,
    pub environment: Option<&'a str>,
}

impl NewGitHubConfig<'_> {
    pub async fn insert(&self, conn: &mut AsyncPgConnection) -> QueryResult<GitHubConfig> {
        self.insert_into(trustpub_configs_github::table)
            .returning(GitHubConfig::as_returning())
            .get_result(conn)
            .await
    }
}

File: packages/crates-io-msw/serializers/trustpub/github-config.ts
import type { components } from '@crates-io/api-client';
import type { TrustpubGithubConfig } from '../../models/index.js';

type ApiGitHubConfig = components['schemas']['GitHubConfig'];

export function serializeGitHubConfig(config: TrustpubGithubConfig): ApiGitHubConfig {
  return {
    id: config.id,
    crate: config.crate.name,
    repository_owner: config.repository_owner,
    repository_owner_id: config.repository_owner_id,
    repository_name: config.repository_name,
    workflow_filename: config.workflow_filename,
    environment: config.environment,
    created_at: config.created_at,
  };
}
